src/analysis/processing/qgsalgorithmlinedensity.cpp

```diff
@@ -0,0 +1,193 @@
+
+#include "qgsalgorithmlinedensity.h"
+#include "qgsgeometryengine.h"
+#include "qgsrasterfilewriter.h"
+
+///@cond PRIVATE
+
+QString QgsLineDensityAlgorithm::name() const
+{
+  return QStringLiteral( "linedensity" );
+}
+
+QString QgsLineDensityAlgorithm::displayName() const
+{
+  return QObject::tr( "Line density" );
+}
+
+QStringList QgsLineDensityAlgorithm::tags() const
+{
+  return QObject::tr( "density,kernel,line,line density,interpolation,weight" ).split( ',' );
+}
+
+QString QgsLineDensityAlgorithm::group() const
+{
+  return QObject::tr( "Interpolation" );
+}
+
+QString QgsLineDensityAlgorithm::groupId() const
+{
+  return QStringLiteral( "interpolation" );
+}
+
+void QgsLineDensityAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input line layer" ), QList<int>() << QgsProcessing::TypeVectorLine ) );
+  addParameter( new QgsProcessingParameterField( QStringLiteral( "WEIGHT" ), QObject::tr( "Weight field " ), QVariant(), QStringLiteral( "INPUT" ), QgsProcessingParameterField::Numeric ) );
+  addParameter( new QgsProcessingParameterDistance( QStringLiteral( "RADIUS" ), QObject::tr( "Search radius" ), 10, QStringLiteral( "INPUT" ), false, 0 ) );
+  addParameter( new QgsProcessingParameterDistance( QStringLiteral( "PIXEL_SIZE" ), QObject::tr( "Pixel size" ), 10, QStringLiteral( "INPUT" ), false ) );
+
+  addParameter( new QgsProcessingParameterRasterDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Line density raster" ) ) );
+}
+
+QString QgsLineDensityAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "The line density interpolation algorithm calculates a density measure of linear features "
+                      "which is obtained in a circular neighborhood within each raster cell. "
+                      "First, the length of the segment of each line that is intersected by the circular neighborhood "
+                      "is multiplied with the lines weight factor. In a second step, all length values are summed and "
+                      "divided by the area of the circular neighborhood. This process is repeated for all raster cells."
+                    );
+}
+
+QgsLineDensityAlgorithm *QgsLineDensityAlgorithm::createInstance() const
+{
+  return new QgsLineDensityAlgorithm();
+}
+
+bool QgsLineDensityAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+  mSource.reset( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+  mWeight = parameterAsString( parameters, QStringLiteral( "WEIGHT" ), context );
+  mSearchRadius = parameterAsDouble( parameters, QStringLiteral( "RADIUS" ), context );
+  mPixelSize = parameterAsDouble( parameters, QStringLiteral( "PIXEL_SIZE" ), context );
```

Comment:
I wondering if an exception or warning should be raised if the pixel size is large compared to the search radius (i.e. if the search radius is completely enclosed by a pixel) -- this would mean some features could potentially fall into gaps between the search circles

Reply:
Yes, I had the same thoughts here... I think I will add an exception that the radius must be at least the square root of `mPixelSize` (=minimum bounding circle of cell).